From 023599a55a8e91ba977a2e1cb9e6ec44b3d8acd6 Mon Sep 17 00:00:00 2001
From: Xiaosong Wei <xiaosong.wei@intel.com>
Date: Mon, 19 Aug 2013 19:17:27 +0800
Subject: color issue fixed in VideoEditor

ColorConvert is only needed if OMX Decoder output color format is NOT
I420.

Category: AOSP Enhancement
Domain: MEDIA
Upstream-Candidate: yes

Change-Id: Iab7f80f12b32b9de252d5feef7fdddb13eae8394
Signed-off-by: Xiaosong Wei <xiaosong.wei@intel.com>
---
 .../src/VideoEditorVideoDecoder.cpp                |   19 +++++++++----------
 1 files changed, 9 insertions(+), 10 deletions(-)

diff --git a/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp b/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
index 417451e..8be9593 100644
--- a/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
+++ b/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
@@ -1050,19 +1050,18 @@ M4OSA_ERR VideoEditorVideoDecoder_create(M4OSA_Context *pContext,
     pDecShellContext->mVideoDecoder->getFormat()->setInt32(kKeyHeight,
         pDecShellContext->m_pVideoStreamhandler->m_videoHeight);
 
-    // Get the color converter
-    pDecShellContext->mI420ColorConverter = new I420ColorConverter;
-    if (pDecShellContext->mI420ColorConverter->isLoaded()) {
-        decoderOutput = pDecShellContext->mI420ColorConverter->getDecoderOutputFormat();
+    if (pDecShellContext->decOuputColorFormat != OMX_COLOR_FormatYUV420Planar) {
+        ALOGV("OMX decoder output format = 0x%X need ColorConverter", pDecShellContext->decOuputColorFormat);
+        pDecShellContext->mI420ColorConverter = new I420ColorConverter;
+        if (pDecShellContext->mI420ColorConverter->isLoaded()) {
+            decoderOutput = pDecShellContext->mI420ColorConverter->getDecoderOutputFormat();
+            ALOGV("decoder output format = 0x%X\n", decoderOutput);
+        }
     }
-
-    if (decoderOutput == OMX_COLOR_FormatYUV420Planar) {
-        delete pDecShellContext->mI420ColorConverter;
-        pDecShellContext->mI420ColorConverter = NULL;
+    else{
+        ALOGV("OMX Decoder already output YUV420Planar, don't need ColorConverter");
     }
 
-    ALOGI("decoder output format = 0x%X\n", decoderOutput);
-
     // Configure the buffer pool from the metadata
     err = VideoEditorVideoDecoder_configureFromMetadata(pDecShellContext,
         pDecShellContext->mVideoDecoder->getFormat().get());
-- 
1.7.4.1

